service-base/src/main/java/org/eclipse/hono/service/monitoring/AbstractMessageSenderConnectionEventProducer.java

```diff
@@ -0,0 +1,112 @@
+/**
+ * Copyright (c) 2018 Contributors to the Eclipse Foundation
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 1.0 which is available at
+ * https://www.eclipse.org/legal/epl-v10.html
+ *
+ * SPDX-License-Identifier: EPL-1.0
+ */
+package org.eclipse.hono.service.monitoring;
+
+import java.nio.charset.StandardCharsets;
+import java.util.Objects;
+import java.util.function.BiFunction;
+
+import org.eclipse.hono.client.HonoClient;
+import org.eclipse.hono.client.MessageSender;
+import org.eclipse.hono.service.auth.device.Device;
+import org.eclipse.hono.util.RegistrationConstants;
+
+import io.vertx.core.CompositeFuture;
+import io.vertx.core.Future;
+import io.vertx.core.json.JsonObject;
+
+/**
+ * A connection event producer based on a {@link MessageSender}.
+ */
+public abstract class AbstractMessageSenderConnectionEventProducer implements ConnectionEventProducer {
+
+    private static final String MESSAGE_CONTENT_TYPE = "application/vnd.eclipse-hono-dc-notification+json";
+
+    private final HonoClient client;
+    private final BiFunction<HonoClient, String, Future<MessageSender>> messageSenderSource;
+
+    protected AbstractMessageSenderConnectionEventProducer(
+            final HonoClient client,
+            final BiFunction<HonoClient, String, Future<MessageSender>> messageSenderSource) {
+
+        Objects.requireNonNull(client);
+        Objects.requireNonNull(messageSenderSource);
+
+        this.client = client;
+        this.messageSenderSource = messageSenderSource;
+    }
+
+    @Override
+    public Future<?> connected(final String remoteId, final String protocolAdapter, final Device authenticatedDevice,
+            final JsonObject data) {
+        return sendNotificationEvent(authenticatedDevice, protocolAdapter, remoteId, "connected", data);
+    }
+
+    @Override
+    public Future<?> disconnected(final String remoteId, final String protocolAdapter,
+            final Device authenticatedDevice, final JsonObject data) {
+        return sendNotificationEvent(authenticatedDevice, protocolAdapter, remoteId, "disconnected", data);
+    }
+
+    private Future<?> sendNotificationEvent(final Device authenticatedDevice, final String protocolAdapter,
+            final String remoteId, final String cause, final JsonObject data) {
+
+        if (authenticatedDevice == null) {
+            // we only handle authenticated devices
+            return Future.succeededFuture();
+        }
+
+        // get an assertion
+
+        final Future<String> assertionFuture = this.client
+                .getOrCreateRegistrationClient(authenticatedDevice.getTenantId())
```

Comment:
this will not work because the Device Registration service will in general not be co-hosted with the Event API implementation. in the default setup the Device Registration service is implemented by the Device Registry whereas the Event API is implemented by Hono Messaging or the Dispatch Router resp.